### DISCUSSION TITLE
ctx.req is undefined in Document

### DISCUSSION THREAD

**Discussion by bluelovers** [vercel/next.js#18235] [Help] [answered]
# Bug report

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

```ts
class MyDocument extends Document
{

	static async getInitialProps(ctx: DocumentContext)
	{
		console.log(`MyDocument`, `getInitialProps`, this)

		if (1 || typeof window === 'undefined')
		{
			const state = getNextGlobalState();
			state.hostInfo = getNextHost(ctx) ?? state.hostInfo;

			const _ctx = getNextPageContext(ctx);

			console.dir(ctx)
			console.dir(_ctx)

			console.dir(state)
		}

		const initialProps = await Document.getInitialProps(ctx);

		return initialProps
	}
```

## Screenshots

![image](https://user-images.githubusercontent.com/167966/97119498-879afe80-174b-11eb-9d62-5b38c2a02918.png)

## System information

- OS: [e.g. macOS, Windows]
- Browser (if applies) [e.g. chrome, safari]
- Version of Next.js: [e.g. 6.0.2]
- Version of Node.js: [e.g. 10.10.0]

**Comment by jamesmosier** [comment, marked as answer]
Hi @bluelovers. `_document.js` is not the recommended place for code like this. Document is not meant to hold application logic like this. There are also some  other caveats](https://nextjs.org/docs/advanced-features/custom-document#caveats) too.

Code like what you have above should be put in [`_app.js`](https://nextjs.org/docs/advanced-features/custom-app). That will achieve what you are looking for and will run on every page.

**Reply by eschaefer** [reply]
There are cases where you need server-side cookies to modify the parent `html` element or `body` element, for example. This must happen in `_document.js` and cannot happen elsewhere.

My workaround was to add an empty `getServerSideProps` call to the appropriate `pages/[pagename].js` file further down the tree. Then `ctx.req` in `_document.js` has the expected headers, cookies, etc. [More background here](https://eric-schaefer.com/til/2022/02/05/next.js-unexpected-missing-ctx.req-in-_document/).
